[18.0] [IMP] l10n_it_delivery_note: create invoice from delivery notes with 1:1 lines correspondence

Borruso opened the pull request:
Migration to 18.0
#4385

odooNextev committed:
[IMP] l10n_it_delivery_note: create invoice from delivery notes

Borruso committed:
[IMP] l10n_it_delivery_note: added configuration checks to handle invoice creation arbitrarily

Borruso committed:
[FIX] l10n_it_delivery_note: fix line display_type in invoice generation

Borruso committed:
[REF] l10n_it_delivery_note: improve invoice generation with line display_type

Borruso committed:
[FIX] l10n_it_delivery_note: correct quantity computation for returns when invoicing from DN

Removed the return check since invoicing is now always performed
from the Delivery Note. Adjusted quantity computation to properly
account for returned quantities.

Borruso committed:
[IMP] l10n_it_delivery_note: avoid error when invoicing delivery notes filtered by multiple partners

Borruso committed:
[REF] l10n_it_delivery_note: refactoring invoice generation and preserving sales lines in invoice lines including notes and sections

Co-authored-by: Babbato <abarbato@dinamicheaziendali.it>

Borruso commented:
@odooNextev ho aggiunto dei nuovi commit se ti va fare review così se ti piace possiamo portarlo anche alla 16.0

stenext commented:
> @odooNextev ho aggiunto dei nuovi commit se ti va fare review così se ti piace possiamo portarlo anche alla 16.0

Ottimo, appena riesco lo faccio

Borruso committed:
[IMP] l10n_it_delivery_note: invoice kits as SO lines instead of DN components

aleboldro02 commented on l10n_it_delivery_note/models/stock_delivery_note.py:
```diff
@@ -667,134 +669,428 @@ def action_confirm(self):
+            #  GROUPING
+            if not grouped:
+                invoice_vals_list = self._group_invoice_vals(
+                    invoice_vals_list, sale_orders
```
Utilizzare filtered_sales invece di sale_orders.

aleboldro02 commented on l10n_it_delivery_note/models/stock_delivery_note.py:
```diff
@@ -667,134 +669,428 @@ def action_confirm(self):
-            for line, vals in cache.items():
-                line.write(vals)
+            # update delivery notes status
+            filter_delivery_notes._update_invoice_statuses(moves, sale_orders)
```
Sostituire sale_orders con filtered_sales

aleboldro02 commented on l10n_it_delivery_note/models/stock_delivery_note.py:
```diff
@@ -667,134 +669,428 @@ def action_confirm(self):
+            delivered_qty = delivered_by_product.get(bom_line.product_id.id, 0.0)
+            ratios.append(delivered_qty / bom_line.product_qty)
+
+        return min(ratios) if ratios else 0.0
```
La fatturazione di un kit relativa alla distinta base potrebbe non essere veritiera nel momento in cui viene cambiata la configurazione del kit durante le varie spedizioni.
Per la tipologia kit, valuterei: il vincolo di fatturazione solo tramite Ordine di Vendita oppure un modello d'appoggio che esegue una fotografia della composizione del kit al momento della conferma dell'OdV o generazione della consegna.
Altrimenti con tale logica variando la distinta base, si potrebbero avere quantità fatturate incoerenti con il kit inizialmente configurato.

Borruso replied:
il problema è se fatturi da ordine comunque avvia la funzione di fattura del DdT come è impostato

```
def _create_invoices(self, grouped=False, final=False, date=None):
        # TODO: Consider adding an 'invoice_policy' selection field on sale.order
        # to allow choosing between 'delivered' (standard) and 'delivery_note' modes.
        # In 'delivery_note' mode:
        # - invoice_status should be computed based on delivery notes, not delivered qty
        # - "Create Invoice" button should appear only when new delivery notes exist
        # This would avoid bypassing standard invoicing when delivery notes exist
        # but user wants to invoice directly (e.g., for multi-delivery orders).
        if any(dn.invoice_status == "to invoice" for dn in self.delivery_note_ids):
            self.delivery_note_ids.action_invoice(
                invoice_method="service", final=final, sale_orders=self, grouped=grouped
            )
            # Assign delivery notes to invoices and update statuses
            self._assign_delivery_notes_invoices(self.invoice_ids.ids)
            return self.invoice_ids
        return super()._create_invoices(grouped=grouped, final=final, date=date)
```

aleboldro02 reviewed (changes requested):
1. Primo fix critico. Impedirebbe la fatturazione e comprometterebbe lo stato di fatturazione.
2. Inserire vincolo per BoM da fatturare già consegnata o valutare il blocco di fatturazione kit da DN o eseguire una fotografia della BoM.
Creare test che coprano tali scenari:
- Fatturazione con termini di pagamento multipli
- Verificare logica acconti con partner multipli.
- Fatturazione di un kit dove la struttura è cambiata.

SirPyTech committed:
[IMP] l10n_it_delivery_note: Group invoice lines by DN

aleboldro02 commented:
@Borruso le due righe commentate in precedenza, dove utilizzi sale_orders sono state volutamente non modificate?
C'è una motivazione?

Borruso commented:
> @Borruso le due righe commentate in precedenza, dove utilizzi sale_orders sono state volutamente non modificate? C'è una motivazione?

mi puoi indicare dove pensavo di averlo cambiato

aleboldro02 commented:
> > @Borruso le due righe commentate in precedenza, dove utilizzi sale_orders sono state volutamente non modificate? C'è una motivazione?
> 
> mi puoi indicare dove pensavo di averlo cambiato

`stock_delivery_note.py` - Linee: 1051-1053 & 1080

Borruso commented:
> `stock_delivery_note.py` - Linee: 1051-1053 & 1080

<img width="589" height="626" alt="Screenshot from 2026-07-24 10-59-39" src="https://github.com/user-attachments/assets/5a66286e-48ec-43e5-a426-322604f14306" />

a riga 1051 non è ancora valorizzata la variabile
1080 c'è `filtered_sales`

aleboldro02 commented:
> a riga 1051 non è ancora valorizzata la variabile 1080 c'è `filtered_sales`

Pardon, non vedevo le sorgenti aggiornate.

aleboldro02 reviewed (approved):
LGTM!

francesco-ooops reviewed (changes requested):
Problema funzionale: la fatturazione da DN sembra fatturare l'intero ordine e non solo il DN selezionato

https://www.loom.com/share/5168085f5ffe4226a59642fede4f0ec3

SirPyTech committed:
[FIX] l10n_it_delivery_note: Invoicing a Return creates a Credit Note

SirPyTech committed:
[IMP] l10n_it_delivery_note: Substitute Maintainer

SirPyTech committed:
[FIX] l10n_it_delivery_note: Invoicing kits

Steps:
1. Create a Kit with 2 Components
2. Sale the Kit and ship it: the components are shipped
3. Create a Delivery Note and invoice it

Before this change:
The Kit is invoiced 2 times

After this change:
The Kit is invoiced once

SirPyTech committed:
[REF] l10n_it_delivery_note: Do not repeat invoiceable lines extraction

francesco-ooops commented:
@Borruso taggami quando è pronta

PS: nelle PR 14 indicate in in https://github.com/OCA/l10n-italy/issues/3979 ci sono dei test per casistiche che abbiamo riscontrato dopo la modifica, magari puoi riciclarli per verificare i comportamenti qui
